src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -0,0 +1,234 @@
+                }
+                else
+                {
+                    enumerateRootPath = workingDirectory;
+                }
+
+                enumeratePattern = Path.GetFileName(rootRule);
+                enumerateDepth = SearchOption.TopDirectoryOnly;
+            }
+            // starts with glob
+            else if(firstGlob == 0)
+            {
+                if(workingDirectory == null) throw new InvalidOperationException();
+                enumerateRootPath = workingDirectory;
+                enumeratePattern = "*";
+                enumerateDepth = SearchOption.AllDirectories;
+            }
+            else
+            {
+                int rootDirLength = rootRule.Substring(0,firstGlob).LastIndexOfAny( new [] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar});
+                enumerateRootPath = rootRule.Substring(0,rootDirLength);
+                enumeratePattern = "*";
+                enumerateDepth = SearchOption.AllDirectories;
+            }
+        }
+
+        public static Fingerprint CreateFingerprint(
+            AgentTaskPluginExecutionContext context,
+            IEnumerable<string> keySegments)
+        {
+            var sha256 = new SHA256Managed();
+
+            string workingDirectoryValue = context.Variables.GetValueOrDefault(
+                "system.defaultworkingdirectory" // Constants.Variables.System.DefaultWorkingDirectory
+                )?.Value;
+
+            var resolvedSegments = new List<string>();
+
+            foreach (string keySegment in keySegments)
+            {
+                bool isPathy = IsPathy(keySegment);
+                bool isWildCard = keySegment.Equals("**", StringComparison.Ordinal);
+
+                if (isWildCard)
+                {
+                    resolvedSegments.Add("**");
+                }
+                else if (isPathy)
+                {
+                    context.Verbose($"Interpretting `{keySegment}` as a path.");
+
+                    var segment = new StringBuilder();
+                    bool foundFile = false;
+
+                    string[] pathSections = keySegment.Split(new []{';'}, StringSplitOptions.RemoveEmptyEntries);
+                    foreach(string pathSection in pathSections)
+                    {
+                        string[] pathRules = pathSection.Split(new []{','}, StringSplitOptions.RemoveEmptyEntries);
+                        string rootRule = pathRules.First();
+                        if(rootRule.Length == 0 || rootRule[1] == '!')
+                        {
+                            throw new ArgumentException();
+                        }
+
+                        string workingDirectory = null;
+                        if (!IsAbsolutePath(rootRule))
+                        {
+                            workingDirectory = workingDirectoryValue;
+                        }
+
+                        string absoluteRootRule = MakePathAbsolute(workingDirectory, rootRule);
```

Comment:
@fadnavistanmay  @jahsu-MSFT @owenhuynMSFT  Is there globbing in DPA we can leverage here?

Reply:
Not that I know of.

Reply:
DPA is doing something like this 
targetPath = Path.IsPathFullyQualified(targetPath) ? targetPath : Path.GetFullPath(Path.Combine(defaultWorkingDirectory, targetPath));